lib/framedivider.rb

```diff
@@ -0,0 +1,305 @@
+
+  pts[:C][:n_prev_unto_next_pl] = pts[:C][:prev_unto_next_pl] - pts[:C][:pt]
+  pts[:C][:n_next_unto_prev_pl] = pts[:C][:next_unto_prev_pl] - pts[:C][:pt]
+
+  pts[:D][:n_prev_unto_next_pl] = pts[:D][:prev_unto_next_pl] - pts[:D][:pt] if four
+  pts[:D][:n_next_unto_prev_pl] = pts[:D][:next_unto_prev_pl] - pts[:D][:pt] if four
+
+  # Fetch angle between both extended vectors (A>pC & A>pB), then normalize (Cn).
+  #
+  #             pC
+  #        eC   ^
+  #          \  |
+  #           \ Cn
+  #            \|
+  # <---------- A <---------- B
+  #             |\
+  #             | \
+  #             |  \
+  #             |   C (or D)
+  #
+  pts[:A][:angle] = pts[:A][:n_prev_unto_next_pl].angle(pts[:A][:n_next_unto_prev_pl])
+  pts[:A][:n_prev_unto_next_pl].normalize!
+  pts[:A][:n_next_unto_prev_pl].normalize!
+
+  pts[:B][:angle] = pts[:B][:n_prev_unto_next_pl].angle(pts[:B][:n_next_unto_prev_pl])
+  pts[:B][:n_prev_unto_next_pl].normalize!
+  pts[:B][:n_next_unto_prev_pl].normalize!
+
+  pts[:C][:angle] = pts[:C][:n_prev_unto_next_pl].angle(pts[:C][:n_next_unto_prev_pl])
+  pts[:C][:n_prev_unto_next_pl].normalize!
+  pts[:C][:n_next_unto_prev_pl].normalize!
+
+  pts[:D][:angle] = pts[:D][:n_prev_unto_next_pl].angle(pts[:D][:n_next_unto_prev_pl]) if four
+  pts[:D][:n_prev_unto_next_pl].normalize! if four
+  pts[:D][:n_next_unto_prev_pl].normalize! if four
+
+  # Generate new 3D points A', B', C' (and D')
+  #
+  #
+  #
+  #
+  #     A' ---------------------- B'
+  #      \
+  #       \      A <---------- B
+  #        \      \
+  #         \      \
+  #          \      \
+  #           C'      C
+  pts[:A][:p] = pts[:A][:pt] + (pts[:A][:n_prev_unto_next_pl] * width)
+  pts[:A][:from_next].normalize!
+  pts[:A][:p] = pts[:A][:p]  + (pts[:A][:from_next] * width * Math.tan(pts[:A][:angle]/2))
+
+  pts[:B][:p] = pts[:B][:pt] + (pts[:B][:n_prev_unto_next_pl] * width)
+  pts[:B][:from_next].normalize!
+  pts[:B][:p] = pts[:B][:p]  + (pts[:B][:from_next] * width * Math.tan(pts[:B][:angle]/2))
+
+  pts[:C][:p] = pts[:C][:pt] + (pts[:C][:n_prev_unto_next_pl] * width)
+  pts[:C][:from_next].normalize!
+  pts[:C][:p] = pts[:C][:p]  + (pts[:C][:from_next] * width * Math.tan(pts[:C][:angle]/2))
+
+  pts[:D][:p] = pts[:D][:pt] + (pts[:D][:n_prev_unto_next_pl] * width) if four
+  pts[:D][:from_next].normalize! if four
+  pts[:D][:p] = pts[:D][:p]  + (pts[:D][:from_next] * width * Math.tan(pts[:D][:angle]/2)) if four
+
+  # Validate if rough opening 'fits' within host surface (think acute angles).
+  # ref: https://github.com/NREL/OpenStudio/blob/
+  #              e603410ac3567d44e7660294f3b06a8a2d591118/ruby/test/
+  #              Polygon3d_PointInPolygonUp_Test.rb
+  vec = OpenStudio::Point3dVector.new
+  dad.each do |vertex|
+    vec.push(OpenStudio::Point3d.new(vertex.x, vertex.y, vertex.z))
+  end
+  # polygon = OpenStudio::Polygon3d.new(vec) <<< error : uninitialized constant OpenStudio::Polygon3d
```

Comment:
@macumber  : unsuccessful initializing Polygon3D (what am I not _requiring_ as a Ruby gem?)